SparseGrids/tsgRuleLocalPolynomial.hpp

```diff
@@ -364,6 +463,79 @@ class templRuleLocalPolynomial : public BaseRuleLocalPolynomial{
         }
         return value;
     }
+
+    // Same logic as the corresponding eval functions.
+    double diffPWQuadratic(int point, double x) const{
+        if (rule == rule_localp) {
+            if (point == 1) return -1.0;
+            if (point == 2) return 1.0;
+        } else if (rule == rule_localpb) {
+            if (point == 0) return -1.0;
+            if (point == 1) return 1.0;
+        }
+        return -2.0 * x;
+    }
+    double diffPWCubic(int point, double x) const{
+        if (rule == rule_localp) {
+            if (point == 0) return 0.0;
+            if (point == 1) return -1.0;
+            if (point == 2) return 1.0;
+            if (point <= 4) return -2.0 * x;
+        } else if (rule == rule_localpb) {
+            if (point == 0) return -1.0;
+            if (point == 1) return 1.0;
+            if (point == 2) return -2.0 * x;
+        } else if (rule == rule_localp0) {
+            if (point == 0) return -2.0 * x;
+        }
+        return (point % 2 == 0) ? 1.0 / 3.0 - x * (x + 2.0) : -1.0 / 3.0 + x * (x - 2.0);
+    }
+    double diffPWPower(int point, double x) const{
+        if (rule == rule_localp)     if (point <= 8) return diffPWCubic(point, x);
+        if (rule == rule_semilocalp) if (point <= 4) return diffPWCubic(point, x);
+        if (rule == rule_localpb)    if (point <= 4) return diffPWCubic(point, x);
+        if (rule == rule_localp0)    if (point <= 2) return diffPWCubic(point, x);
+        int level = getLevel(point);
+        int max_ancestors;
+        if (rule == rule_localp)     max_ancestors = level-2;
+        if (rule == rule_semilocalp) max_ancestors = level-1;
+        if (rule == rule_localpb)    max_ancestors = level-1;
+        if (rule == rule_localp0)    max_ancestors = level;
+        if (max_order > 0) max_ancestors = std::min(max_ancestors, max_order - 2);
+        int most_turns = 1;
+        double phantom_distance = 1.0;
+        auto update_and_get_next_node = [&]() {
+            most_turns *= 2;
+            phantom_distance = 2.0 * phantom_distance + 1.0;
+            int turns = (rule == rule_localp0) ? ((point+1) % most_turns) : ((point-1) % most_turns);
+            return (turns < most_turns / 2) ?
+                    (phantom_distance - 2.0 * ((double) turns)) :
+                    (-phantom_distance + 2.0 * ((double) (most_turns - 1 - turns)));
+        };
+
+        // Does not include the additional factor (1-x)(1+x) and the Lagrange coefficient.
+        std::vector<double> nodes(max_ancestors), left_prods(max_ancestors);
```

Comment:
Is there a way to avoid allocating/deallocating vectors here?

Reply:
I avoided allocating `nodes` by using another lambda that inverts the operations made by `update_and_get_next_node()`, but I don't think you can avoid allocating `left_prods()` unless we allocate it outside of this function and pass it in as an input.